azure/Guidelines.md

```diff
@@ -747,6 +749,34 @@ While removing a value from an enum is a breaking change, adding value to an enu
 
 > :ballot_box_with_check: **You SHOULD** use extensible enums unless you are positive that the symbol set will **NEVER** change over time.
 
+### Deprecating Behavior Notification
+
+When the [API Versioning](#API-Versioning) guidance above cannot be followed and the [Azure Breaking Change Reviewers](mailto:azbreakchangereview@microsoft.com) approve a [breaking change](#123-definition-of-a-breaking-change) to a specific API version it must be communicated to its callers. The API version that is being deprecated must add the `azure-deprecating` response header with a semicolon-delimited string notifying the caller what is being deprecated, when it will no longer function, and a URL linking to more information such as what new operation they should use instead.
+
+The purpose is to inform customers (when debugging/logging responses) that they must take action to modify their call to the service's operation and use a newer API version or their call will soon stop working entirely. It is not expected that client code will examine/parse this header's value in any way; it is purely informational to a human being. The string is _not_ part of an API contract (except for the semi-colon delimiters) and may be changed/improved at any time without incurring a breaking change.
```

Comment:
Whilst deprecation times may vary on a service-by-service basis, it'd be good to have some centralised guidelines for the minimum deprecation notice period here.

We've had a number of awkward breaking changes over the years (e.g. Soft Delete for Key Vault being enforced on December 31st) and deprecations ([service teams yanking API versions with (AFAICT) no heads up](https://github.com/hashicorp/terraform-provider-azurerm/issues/18185)) - that having a centralised minimum deprecation period (e.g. 6 months?) would be useful/worth having here?

Specifically around the Key Vault Soft Delete changes, would it also be worth recommending avoiding deprecation periods in the times of year where users are likely to be less available (e.g. don't schedule them in christmas/new year/the summer)?

Reply:
Officially, we do: When Azure announces a deprecation, the currently-shipping versions should work as-is for at least 3 years. This shrinks to 1 year [possibly less] if the old versions have a security or compliance issue and we need to get customers to move away from them more aggressively). We do batch the deprecation announcements so that they come out just a few times per year. We should be able to announce anytime (including x-mas/new year) because the old api-versions should continue running as-is for 3/1 years).

Occassionally, teams introduce breaking behaviors by mistake as our engineers are human, but we correct these as soon as we discover/are notified of them.

Reply:
That's good to know! Is this published anywhere that we can link to if we need to refer to that guideline?